lib/spack/spack/filesystem_view.py

```diff
@@ -0,0 +1,524 @@
+            should be removed as well.
+
+            Should except an `exclude` keyword argument containing a list of
+            regexps that filter out matching spec names.
+
+            This method should make use of `deactivate_{extension,standalone}`.
+        """
+        raise NotImplementedError
+
+    def remove_extension(self, spec):
+        """
+            Remove (unlink) an extension from this view.
+        """
+        raise NotImplementedError
+
+    def remove_standalone(self, spec):
+        """
+            Remove (unlink) a standalone package from this view.
+        """
+        raise NotImplementedError
+
+    def get_all_specs(self):
+        """
+            Get all specs currently active in this view.
+        """
+        raise NotImplementedError
+
+    def get_spec(self, spec):
+        """
+            Return the actual spec linked in this view (i.e. do not look it up
+            in the database by name).
+
+            `spec` can be a name or a spec from which the name is extracted.
+
+            As there can only be a single version active for any spec the name
+            is enough to identify the spec in the view.
+
+            If no spec is present, returns None.
+        """
+        raise NotImplementedError
+
+    def print_status(self, *specs, **kwargs):
+        """
+            Print a short summary about the given specs, detailing whether..
+                * ..they are active in the view.
+                * ..they are active but the activated version differs.
+                * ..they are not activte in the view.
+
+            Takes `with_dependencies` keyword argument so that the status of
+            dependencies is printed as well.
+        """
+        raise NotImplementedError
+
+
+class YamlFilesystemView(FilesystemView):
+    """
+        Filesystem view to work with a yaml based directory layout.
+    """
+
+    def __init__(self, root, layout, **kwargs):
+        super(YamlFilesystemView, self).__init__(root, layout, **kwargs)
+
+        self.extensions_layout = YamlViewExtensionsLayout(root, layout)
+
+        self._croot = colorize_root(self.root) + " "
+
+    def add_specs(self, *specs, **kwargs):
+        assert all((s.concrete for s in specs))
+        specs = set(specs)
+
+        if kwargs.get("with_dependencies", True):
+            specs.update(get_dependencies(specs))
+
+        if kwargs.get("exclude", None):
+            specs = set(filter_exclude(specs, kwargs["exclude"]))
+
+        conflicts = self.get_conflicts(*specs)
+
+        if conflicts:
+            for s, v in conflicts:
+                self.print_conflict(v, s)
+            return
+
+        extensions = set(filter(lambda s: s.package.is_extension, specs))
+        standalones = specs - extensions
+
+        set(map(self._check_no_ext_conflicts, extensions))
+        # fail on first error, otherwise link extensions as well
+        if all(map(self.add_standalone, standalones)):
```

Comment:
I'm starting to look at this now. I noticed if one of the packages is external and has a prefix like:

```
packages:
    openssl:
        paths:
            openssl@1.0.1e: /usr/local
```

That all of `/usr/local` will be added. Could you add a check to skip external specs here?

Reply:
Ew, yeah, I'll do that.

Reply:
OK, fixed and I've added a test for it.